### DISCUSSION TITLE
Expected behavior of subplot frame annotations

### DISCUSSION THREAD

**Discussion by PaulWessel** [GenericMappingTools/gmt#4756] [Code development discussion]
Given we currently have some minor differences between the behavior in master and gmt-themes branch, I wanted to clarify what I think is the desired behavior before we hunt for bugs.  Below is what I think was the intent (which may differ from current implementation).

There are four distinct cases depending on the use of **-S**:

**1: No -S:**

- Each panel's frames are controlled by the **MAP_FRAME_AXES** setting
- This can be overridden for all panels by adding **-B**_frameargs_ to the **subplot begin** command
- It can further be overridden by adding **-B**_frameargs_ to the first plot command in a panel

**2: Using -SC:**

Each column of panels share the same xmin/xmax range and thus can share axes:

- User chooses to annotate the top and bottom x-axes, or just the top (**-SCt**) or bottom (**-SCb**) axis will be annotated.
- All **other** x-axes shall **not** be annotated and their presence and attributes as a non-annotated axis is under the control of **MAP_FRAME_AXES**.
- That setting can be overridden by adding **-B**_frameargs_ to the to the subplot begin command
- Because of layout settings dictated by **-SC**, **-B**_frameargs_  for panels is not considered.

**3: Using -SR:**

Each row of panels share the same ymin/ymax range and thus can share axes:

- User chooses to annotate the leftmost and rightmost y-axes, or just the top (**-SRl)** or bottom (**-SRr**) axis will be annotated.
- All **other** y-axis are **not** to annotated and their presence and attributes as a non-annotated axis is under the control of **MAP_FRAME_AXES**.
- That setting can be overridden by adding **-B**_frameargs_ to the to the subplot begin command
- Because of layout settings, **-B**_frameargs_  for panels is not considered.

**4: Using both **-SC** and **-SR**

For each of these cases I see four possible variations:

1. Accept the default behavior
2. Override MAP_FRAME_AXES on the gmt subplot begin command
3. Override via -Bframe on the gmt subplot begin command
4. Override via -Bframe for individual panels

This yields a total of 16 demonstrations of 2x2 subplot behaviors.  I have grouped each subplot category so we can examine each case and its four variants.  Note: In master, the default MAP_FRAME_AXES remains WESNZ at this writing so I modify that in the examples.  I realize I am not testing _lrbt_ vs _wesn_ below but we will since modern mode has a WrStZ default. Note: The panel override is always for the lower left (LL) panel in all examples.

**1: No -S:** I think all variants are correct.

![subplot_1](https://user-images.githubusercontent.com/26473567/107100824-8b620680-67b9-11eb-8846-f6dccf01da4b.png)

**2: Using -SC:** I think 2.0 is not correct. The three x-axes outside -SC annotation control should be drawn per WE**SN**. 2.1 works for rows but one x-axis drawn missing (should be like 2.2). 2.3 panel OK but otherwise same issue as 2.0

![subplot_2](https://user-images.githubusercontent.com/26473567/107100811-843af880-67b9-11eb-8703-5cf708d2cc90.png)

**3: Using -SR:** Same problems as case 2 but transposed.

![subplot_3](https://user-images.githubusercontent.com/26473567/107100805-7eddae00-67b9-11eb-8fb7-b0aa95cb7aa9.png)

**4: Using both **-SC** and **-SR** CLoser, but 4.1 shows MAP_FRAME_AXES is not consulted.  The rest looks OK

![subplot_4](https://user-images.githubusercontent.com/26473567/107100771-6cfc0b00-67b9-11eb-9417-6feee8ad27db.png)

**Comment by PaulWessel** [comment]
I think I should add these scripts to our modern test suite (all passing for now until we make any changes).  Pinging @GenericMappingTools/core and @GenericMappingTools/gmt-contributors to comment on this discussion item.

**Reply by maxrjones** [reply]
Does the behavior in 2.0 and 3.0 match your expectation if **+w** is appended? From reading the subplot **-S** [documentation](https://docs.generic-mapping-tools.org/dev/subplot.html#s), my interpretation is that the plotted results are consistent with the documentation, specifically "Append +w to draw horizontal and vertical lines between interior panels using selected pen [no lines]".

**Reply by PaulWessel** [reply]
That **+w** goes with **-F** and was used above to draw the red lines for 2x2 subplots consisting of the PNGs from the 4 smaller 2x2 subplots.  So it has nothing to do with **-S**.

**Reply by maxrjones** [reply]
I understand and agree with your interpretation of the 'correct' behavior.

**Comment by PaulWessel** [comment]
Do @joa-quim and @seisman agree with the above so that I can work on addressing the mistakes identified in master before we try to merge in gmt-themes?

**Reply by seisman** [reply]
I agree with what you said.

**Comment by joa-quim** [comment]
Didn't follow them all, but go.

**Comment by PaulWessel** [comment]
Just for the record, I was not accurate above.  The **MAP_FRAME_AXES** default in master at this time is _WESNZ_.  That means if **-SC** or **-SR** determine which axes should be annotated, the others should be relegated to the lower-case version of what **MAP_FRAME_AXES** says, i.e., _wesn_.  Given that, these are the only one that passes: 1.x (all of the ones with no **-S**), x.2 (all LL panels are correct). The rest fail.
